### DISCUSSION TITLE
Pre-loading custom views

### DISCUSSION THREAD

**Discussion by lukaszossowski** [AvaloniaUI/Avalonia#10384] [Q&A] [answered]
I noticed that on my Avalonia apps that there is a common occurrence - when a view is loaded for the first time, it takes longer than the next time it is created. I am using a content control with datatemplates displaying views like so:

```
<ContentControl Grid.Row="2" Content="{Binding CurrentPageVM}" Margin="0,10,0,0" >
				<ContentControl.DataTemplates>
					<DataTemplate DataType="backend:ComponentPageViewModel">
						<local:ComponentPage />
					</DataTemplate>
				</ContentControl.DataTemplates>
				<ContentControl.DataTemplates>
					<DataTemplate DataType="backend:ProductPageViewModel">
						<local:ProductPage />
					</DataTemplate>
				</ContentControl.DataTemplates>
				<ContentControl.DataTemplates>
					<DataTemplate DataType="backend:InvoicePageViewModel" >
						<local:SalesPage />
					</DataTemplate>
				</ContentControl.DataTemplates>
</ContentControl>
```

I'm guessing that after they are shown for the first time they are cached somehow or something like that, so would it be possible to pre-load all views so that they load quickly all the time?

It is far from optimal for the user to have the app stutter once just at the beginning just to work fine later on.

**Comment by timunie** [comment]
I wonder how complex your view is? I don't see such issues in my App. Other than that, if you are loading data in your ViewModel, make sure that any I/O task is async if that may take some time.

**Reply by lukaszossowski** [reply]
Each time I recreate the viewmodel, so its creation time should be a constant, also while testing there isn't very much going on in terms of loading stuff yet. I do notice the same in my pretty simple UI app as well in a one with quite complex interface (though again only a dozen ro so bound single value properties).

It is also the same with opening new windows - the first time it is opened takes noticably longer. To check it I have written the following helper class:

```
 public static class PageLoadedInformer
    {
        private static Stopwatch Counter;

        public static void PageLoaded()
        {
            if (Counter != null)
            {
                Counter.Stop();
                Trace.WriteLine("Loaded in: " + Counter.ElapsedMilliseconds + "ms");
            }
        }
        public static void StartCounting()
        {
            Counter = Stopwatch.StartNew();
        }
    }
```

I open a window like this:
```

 private async Task<T> OpenWindow<T>(Window window)
        {            
            if (Avalonia.Application.Current?.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop && desktop != null && desktop.MainWindow != null)
            {
                PageLoadedInformer.StartCounting();
                return await window.ShowDialog<T>(desktop.MainWindow);
            }

            throw new Exception("Not on desktop!");
        }
```

And in the window's Loaded event I invoke the PageLoaded() method.

The result looks like this after opening the window 3 times:

Loaded in: 149ms
Loaded in: 13ms
Loaded in: 14ms

This is with some controls. I also tried this as cleanly as possible, just opening the window, without setting the DataContext, window looks like this:
```

<Border Background="{StaticResource OutlineBrush}" CornerRadius="10" BoxShadow="0 0 5 2 #11000000" Margin="5" >
		<Border Background="{StaticResource MainBackgroundBrush}" CornerRadius="9" Margin="1" ClipToBounds="True">
			<Grid  RowDefinitions="Auto,10,*,10,40" ColumnDefinitions="*,10,*" Margin="10" >
				<Grid Grid.Row="11" ColumnDefinitions="*,10,*" Grid.ColumnSpan="3">

					<Button Name="CancelButton" Content="Cancel" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" VerticalContentAlignment="Center" HorizontalContentAlignment="Center"/>

					<Button Name="AcceptButton" Content="Accept" Grid.Column="2" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" VerticalContentAlignment="Center" HorizontalContentAlignment="Center"/>

				</Grid>
			</Grid>
		</Border>
	</Border>
```

![empty window](https://user-images.githubusercontent.com/112421929/219948138-87144520-386a-4e09-afb6-86888ddae5ac.JPG)

So an empty window with two buttons, no custom controls or anything. The timing results:

Loaded in: 23ms
Loaded in: 6ms
Loaded in: 4ms

So even though in this case the difference is indeed not noticable, but it is definitely there. This is why I assumed that there is something going on that causes it to load faster after initial load. The same happens with complex views etc. but it just becomes a little stutter, especially problematic on mobile if there is an animation being started at the same time.

**Reply by timunie** [reply]
One last question: which version of Avalonia do you use? On mobile there was a huge improvement going on between 0.10 and 11.0 preview packages.

**Reply by lukaszossowski** [reply]
I am on 11.0.0 - preview 4 on all my projects currently.

But I guess that mobile performance is a different issue, not really related to this one - why the first time a window/view loads it is a couple times slower than subsequent ones.

**Comment by ahopper** [comment, marked as answer]
Publishing with ReadyToRun can help with this.

**Reply by lukaszossowski** [reply]
Thank you. This does work and now the bahaviour is consistent.
